### PR TITLE
4512 invalid badblock

### PR DESCRIPTION
**Pull request by jflo**
## PR description

BlockValidationResults and BlockProcessingResults are top level classes with explicit fields to describe failure situations, including the causing exception if any.  This allows one class as a return type which can track validation and processing issues separately.

Adds a convenience method on BlockProcessingOutputs for an empty yield.
Adds messaging or root causes for failures in a few places, shedding light on failures.
New test suite that simulates low level errors at various steps of block import, spanning validation, processing, and output validation.

PR does not separate validation from processing. This should be done for readability and separation of concerns, however was not strictly necessary to make error handling clearer.

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/BlockProcessingResult.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import org.hyperledger.besu.ethereum.core.TransactionReceipt;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class BlockProcessingResult extends BlockValidationResult {
+
+  private final Optional<BlockProcessingOutputs> yield;
+  private final boolean isPartial;
```
Comment (jflo) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/BlockProcessingResult.java:28]: There are rollup usecases that will want processing results from a failed block, including only the successful results.

Comment (garyschulte) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/BlockProcessingResult.java:28]: only feedback here is that isPartial might imply multiple causes and messages.  non-blocking just an observation

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java [hunk 2]
```diff
@@ -148,12 +150,13 @@ public Result processBlock(
...
-              blockHeader.getHash().toHexString(),
-              transaction.getHash().toHexString());
-          return AbstractBlockProcessor.Result.failed();
+          String errorMessage =
+              MessageFormat.format(
+                  "Block processing error: transaction invalid '{}'. Block {} Transaction {}",
+                  result.getValidationResult().getInvalidReason(),
+                  blockHeader.getHash().toHexString(),
+                  transaction.getHash().toHexString());
+          return new BlockProcessingResult(BlockProcessingOutputs.empty(), errorMessage);
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/goquorum/GoQuorumBlockProcessor.java:159]: did you miss a LOG.info(errorMessage) here?

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/MainnetBlockValidatorTest.java [hunk 3]
```diff
@@ -182,25 +160,7 @@ public void shouldDetectAndCacheInvalidBlocksWhenBodyInvalid() {
...
-              public List<TransactionReceipt> getPrivateReceipts() {
-                return Collections.EMPTY_LIST;
-              }
-
-              @Override
-              public boolean isSuccessful() {
-                return true;
-              }
-            });
+        .thenReturn(new BlockProcessingResult(BlockProcessingOutputs.empty()));
```
Comment (garyschulte) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/MainnetBlockValidatorTest.java:163]: 🔥

**Review by garyschulte — APPROVED** [review]
Love it.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java [hunk 4]
```diff
@@ -131,7 +131,7 @@ public BlockProcessingResult validateAndProcessBlock(
         if (!blockBodyValidator.validateBody(
             context, block, receipts, worldState.rootHash(), ommerValidationMode)) {
           handleAndLogImportFailure(block, result);
-          return result;
+          return new BlockProcessingResult("failed to validate output of imported block");
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java:134]: 👍

**Pull request merged**